DM-38916: Link to bind documentation from query methods

kfindeisen opened the pull request:
This PR adds cross-links from the `bind` parameter of `Registry.query*` methods to the general query guide. Most of the diff is fixing more general documentation issues found in the process.

kfindeisen committed:
Fix broken links in Registry documentation.

Many type links didn't work, either because they weren't properly
qualified or because there was no such type in the first place. I also
fixed some formatting errors that confused the parser (in particular,
not putting a space between a parameter and the following colon).

kfindeisen committed:
Document bind parameter of Registry.query*.

The bind parameter has some behaviors that aren't obvious from the
original 1-sentence description. These are documented in the querying
guide; providing a link to there should be enough to make them
visible.

timj commented:
See also the same fixes for the sphinx errors on #840. I guess it depends which PR merges first.

kfindeisen commented:
Oh, sorry. I thought you said that you only got duplication warnings (which I've ignored here).

timj commented:
Sorry. I said that I fixed all the non-duplication warnings yesterday. Your PR has many more fixes than those warnings but there is some duplication.

kfindeisen committed:
Fix corrupted parameter types in docstrings.

I've also fixed some broken links on the affected lines, but did not
trawl the entire file for them.

timj commented on python/lsst/daf/butler/registry/_registry.py:
```diff
@@ -203,19 +204,19 @@ def fromConfig(
         config : `ButlerConfig`, `RegistryConfig`, `Config` or `str`
             Registry configuration
-        butlerRoot : `lsst.resources.ResourcePathExpression`, optional
+        butlerRoot : convertible to `lsst.resources.ResourcePath`, optional
```
There are many places where we refer to `ResourcePathExpression` so I wonder about this change but I see there is precedent in this file.

kfindeisen replied:
Yes, there are, but [the link can't work](https://pipelines.lsst.io/v/d_2023_06_05/py-api/lsst.daf.butler.registry.Registry.html#lsst.daf.butler.registry.Registry.fromConfig) because `ResourcePathExpression` is not a documented type. The same goes for any other Mypy-helper types.

It's not obvious to me, at least, what does or doesn't count as a "ResourcePath expression", so this is one case where I think it makes sense for the human-readable description to deviate from the machine-readable one. I'll admit `DataId` is a case where the meaning should be clear without explanation.

~EDIT: hmm, looks like I missed the `ButlerConfig` and `Config` links~ 😕

kfindeisen replied:
Interesting: it looks like there are *two* `Registry` pages, https://pipelines.lsst.io/v/d_2023_06_05/py-api/lsst.daf.butler.registry.Registry.html and https://pipelines.lsst.io/v/d_2023_06_05/py-api/lsst.daf.butler.Registry.html, and the links work on one and are broken on the other. A good argument for avoiding unqualified links. 😬

timj replied:
Yes, this is one of the things that Sphinx complains about.

> docstring of lsst.daf.butler.registry._registry.Registry:1: WARNING: duplicate object description of lsst.daf.butler.registry._registry.Registry, other instance in py-api/lsst.daf.butler.Registry, use :noindex: for one of them

timj commented on python/lsst/daf/butler/registry/_registry.py:
```diff
@@ -614,7 +614,7 @@ def removeDatasetType(self, name: str | tuple[str, ...]) -> None:
         Parameters
         ----------
-        name : `str` or `tuple[str, ...]`
+        name : `str` or `tuple` [`str`]
```
The ellipsis is important. It says a tuple with any number of strings, versus a tuple of a single string.

kfindeisen replied:
Except that the "tuple" is logically being used as a list, not as a tuple, and [the convention for lists](https://developer.lsst.io/python/numpydoc.html#sequence-types) is pretty clear.

(IMO this method should take any iterable, but that's a code issue, not a documentation formatting issue.)

timj commented on python/lsst/daf/butler/registry/_registry.py:
```diff
@@ -870,7 +870,7 @@ def getDataset(self, id: DatasetId) -> Optional[DatasetRef]:
         Parameters
         ----------
-        id : `DatasetId`
+        id : `uuid.UUID`
```
Formally this really should be a `lsst.daf.butler.DatasetId` which is a type alias for UUID. It seems that Sphinx handles the aliasing just fine in the formal type annotations but gets confused in the numpydoc and doesn't link it properly.

@TallJimbo are we formally deprecating `DatasetId` or do we want to continue to use it as a public-facing alias?

andy-slac replied:
Personally I'd prefer to have an abstraction with the name `DatasetId`. And I think that having an actual type for it would be even better, we can then have all useful methods (e.g. factories) in that class. UUID will likely to stay as the only implementation for it, but our clients do not need to care about that detail.

kfindeisen replied:
FTR, https://pipelines.lsst.io/v/d_2023_06_05/modules/lsst.daf.butler/index.html#classes includes `DatasetId` in its class list, but doesn't have a valid anchor for it (Sphinx warns about this as well; I'd assumed it couldn't be fixed short of removing all references to `DatasetId`). If the documentation of `DatasetId` could be fixed so that it's linkable, then I'd be fine with leaving these references as-is.

timj replied:
Ok. Leave it as it was originally and we can try to fix DatasetId as @andy-slac describes above. There is a doc string attached to `DatasetId` in `ref.py` but clearly that doesn't do anything useful.

TallJimbo replied:
FWIW, I'm more in the "UUID is going to the only implementation forever and we should just use that type directly" camp.  But I certainly agree this isn't the ticket for it.

timj commented on python/lsst/daf/butler/registry/_registry.py:
```diff
@@ -1385,27 +1388,27 @@ def queryDatasets(
             include dimension records (`DataCoordinate.hasRecords` will be
-            `False`) unless `~queries.DatasetQueryResults.expanded` is called
-            on the result object (which returns a new one).
+            `False`) unless `~registry.queries.DatasetQueryResults.expanded` is
```
```suggestion
            `False`) unless `~.queries.DatasetQueryResults.expanded` is
```

This seems to work for me without having to put the `registry` in the path. Similarly above on line 1391.

timj commented on python/lsst/daf/butler/registry/_registry.py:
```diff
@@ -1500,30 +1506,31 @@ def queryDataIds(
         Returns
         -------
-        dataIds : `queries.DataCoordinateQueryResults`
+        dataIds : `registry.queries.DataCoordinateQueryResults`
```
```suggestion
        dataIds : `.queries.DataCoordinateQueryResults`
```
and other places.

timj commented on python/lsst/daf/butler/registry/_registry.py:
```diff
@@ -1658,17 +1668,18 @@ def queryDatasetAssociations(
     def obsCoreTableManager(self) -> ObsCoreTableManager | None:
-        """ObsCore manager instance for this registry (`ObsCoreTableManager`
+        """ObsCore manager instance for this registry
+        (`~lsst.daf.butler.registry.interfaces.ObsCoreTableManager`
```
```suggestion
        (`~.interfaces.ObsCoreTableManager`
```

timj reviewed (approved):
Thanks for doing this. I'd like @TallJimbo or @andy-slac to comment on the `DatasetId` issue as I'm not completely sure we are ready to drop that from the public APIs and replace it with an explicit UUID.
